codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/GoJmespathExpressionGenerator.java

```diff
@@ -135,8 +345,71 @@ private Result visitKeysFunction(List<JmespathExpression> args, Shape current) {
                     v$1L = append(v$1L, k)
                 }""", idIndex, arg.ident);
 
-        return new Result(listOf(STRING_SHAPE), "v" + idIndex);
+        return new Variable(listOf(STRING_SHAPE), "v" + idIndex, sliceOf(GoUniverseTypes.String));
+    }
+
+    private String nextIdent() {
+        ++idIndex;
+        return "v" + idIndex;
+    }
+
+    private Shape expectMember(CollectionShape shape) {
+        return switch (shape.getMember().getTarget().toString()) {
+            case "smithy.go.synthetic#StringList" -> listOf(STRING_SHAPE);
+            case "smithy.go.synthetic#IntegerList" -> listOf(INT_SHAPE);
+            case "smithy.go.synthetic#BooleanList" -> listOf(BOOL_SHAPE);
+            default -> ShapeUtil.expectMember(ctx.model(), shape);
+        };
     }
 
-    public record Result(Shape shape, String ident) {}
+    private Shape expectMember(MapShape shape) {
```

Comment:
Discussed offline. When I first looked at it, I thought `MapShape` was somehow mapping to a List internally. However, this is mapping the target of the shape, and if it's a list (like `StringList`) it will map it to a list, which makes sense.

Disregard this :)